command/debug.go

```diff
@@ -0,0 +1,945 @@
+	var metricsCollection []map[string]interface{}
+	var hostInfoCollection []map[string]interface{}
+
+	// Start a goroutine to collect data as soon as it's received
+	collectCh := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case metricsEntry := <-c.metricsCh:
+				metricsCollection = append(metricsCollection, metricsEntry)
+			case serverStatusEntry := <-c.serverStatusCh:
+				serverStatusCollection = append(serverStatusCollection, serverStatusEntry)
+			case replicationStatusEntry := <-c.replicationStatusCh:
+				replicationStatusCollection = append(replicationStatusCollection, replicationStatusEntry)
+			case hostInfoEntry := <-c.hostInfoCh:
+				hostInfoCollection = append(hostInfoCollection, hostInfoEntry)
+			case <-c.doneCh:
+				close(collectCh)
+				return
+			}
+		}
+	}()
+
+	// Start capture by capturing the first interval before we hit the first
+	// ticker.
+	c.pollingWg.Add(1)
+	go c.intervalCapture(client, idxCount, startTime, errCh)
+
+	c.pollingWg.Add(1)
+	go c.metricsIntervalCapture(client, mIdxCount, errCh)
+
+	// Capture at each interval, until end of duration or interrupt.
+	intervalTicker := time.Tick(c.flagInterval)
+	metricsIntervalTicker := time.Tick(c.flagMetricsInterval)
+POLLING:
+	for {
+		select {
+		case err := <-errCh:
+			c.debugIndex.Errors = append(c.debugIndex.Errors, err)
+		case <-intervalTicker:
+			idxCount++
+			c.pollingWg.Add(1)
+			go c.intervalCapture(client, idxCount, startTime, errCh)
+		case <-metricsIntervalTicker:
+			mIdxCount++
+			c.pollingWg.Add(1)
+			go c.metricsIntervalCapture(client, mIdxCount, errCh)
+		case <-durationCh:
+			// Wait for polling requests to finish before breaking
+			c.pollingWg.Wait()
+
+			break POLLING
+		case <-c.ShutdownCh:
+			c.UI.Info("Caught interrupt signal, exiting...")
+			break POLLING
+		}
+	}
+
+	// Close the done channel to signal termination, make sure collection
+	// goroutine is terminated before proceeding to persisting the info.
+	close(c.doneCh)
+	<-collectCh
+
+	// Write collected data to their corresponding files
+	if err := c.persistCollection(metricsCollection, "metrics.json"); err != nil {
+		c.UI.Error(fmt.Sprintf("Error writing data to %s: %v", "metrics.json", err))
+	}
+	if err := c.persistCollection(serverStatusCollection, "server_status.json"); err != nil {
+		c.UI.Error(fmt.Sprintf("Error writing data to %s: %v", "server_status.json", err))
+	}
+	if err := c.persistCollection(replicationStatusCollection, "replication_status.json"); err != nil {
+		c.UI.Error(fmt.Sprintf("Error writing data to %s: %v", "replication_status.json", err))
+	}
+	if err := c.persistCollection(hostInfoCollection, "host_info.json"); err != nil {
+		c.UI.Error(fmt.Sprintf("Error writing data to %s: %v", "host_info.json", err))
+	}
+
+	return nil
+}
+
+func (c *DebugCommand) intervalCapture(client *api.Client, idxCount int, startTime time.Time, errCh chan<- *captureError) {
+	defer c.pollingWg.Done()
+
+	var wg sync.WaitGroup
```

Comment:
Right now you're forcing all captures to go at the speed of the slowest.  It might be nicer to decouple them so that e.g. a pprof that lags doesn't impede regular collection of replication-status.

Reply:
Each target runs on it's own goroutine, so pprof lag shouldn't impede collection of another target such as replication-status. The waitgroup here simply waits for all targets for this particular interval to be done before returning since `intervalCapture` itself runs on a goroutine.

Reply:
Suppose the interval is 5s, and pprof collection takes 7s.  That means that over 30s we'll get 4 data points for every data source.  If they were decoupled, we would get 4 data points for pprof and 6 data points for the other data sources, assuming they're not similarly slow.

Reply:
The ticker would have ended after the 4th data point (30th second) since the for loop would have been broken by the durationCh case. There wouldn't be a 5th and 6th goroutine running for intervalCapture even if pprof is slow.

Reply:
You're talking about the code as is.  I'm saying that if you broke the coupling like I'm suggesting, there would be a 5th and 6th datapoint for everyone who wasn't slow and was able to keep up with the 5s polling interval.

Reply:
I am not sure if the intent to capture an extra interval is good here. It could lead to a greedy case where two different slow targets ends up extending one another.

Reply:
I tend to agree that each of the capture blocks would model really well as a `oklog/run.Group` actor and would alleviate this concern that nick brought up (which i share). Each capture could add onto a run group and have their own for loop and ticker to kickoff captures.